test/distributed/checkpoint/test_dtensor_resharding.py

```diff
@@ -0,0 +1,253 @@
+        )
+        self.assertEqual(
+            state_dict_to_save["dtensor"].to_local(),
+            state_dict_to_load["dtensor"].to_local(),
+        )
+
+    @with_comms
+    @skip_if_lt_x_gpu(2)
+    @parametrize("two_d_to_two_d_placements", TWO_D_TO_TWO_D_PLACEMENTS)
+    @with_temp_dir
+    def test_2d_to_2d_reshard_placement_change(self, two_d_to_two_d_placements) -> None:
+        CHECKPOINT_DIR = self.temp_dir
+        original_placement, new_placement = two_d_to_two_d_placements
+
+        global_tensor = torch.arange(16, dtype=torch.float).view(4, 4)
+        mesh_shape = (2, self.world_size // 2)
+        mesh_2d = init_device_mesh(self.device_type, mesh_shape)
+        dtensor = distribute_tensor(
+            global_tensor,
+            mesh_2d,
+            placements=original_placement,
+        )
+        state_dict_to_save = {"dtensor": dtensor}
+
+        dist_cp.save_state_dict(
+            state_dict=state_dict_to_save,
+            storage_writer=dist_cp.FileSystemWriter(path=CHECKPOINT_DIR),
+            planner=dist_cp.DefaultSavePlanner(),
+        )
+
+        zero_dtensor = zeros([4, 4], device_mesh=mesh_2d, placements=new_placement)
+        state_dict_to_load = {"dtensor": zero_dtensor}
+
+        dist_cp.load_state_dict(
+            state_dict=state_dict_to_load,
+            storage_reader=dist_cp.FileSystemReader(CHECKPOINT_DIR),
+            planner=dist_cp.DefaultLoadPlanner(),
+        )
+
+        state_dict_to_load["dtensor"] = state_dict_to_load["dtensor"].redistribute(
+            mesh_2d,
+            placements=[Replicate(), Replicate()],
+        )
+        self.assertEqual(global_tensor, state_dict_to_load["dtensor"].to_local())
+
+        state_dict_to_load["dtensor"] = state_dict_to_load["dtensor"].redistribute(
+            mesh_2d,
+            placements=original_placement,
+        )
+        self.assertEqual(
+            state_dict_to_save["dtensor"].to_local(),
+            state_dict_to_load["dtensor"].to_local(),
+        )
+
+
+class TestDTensorReshardMeshChange(DTensorTestBase):
+    """
+    Test DCP reshard for DTensor with placements changes and mesh_tensor change.
+    """
+
+    @with_comms
+    @with_temp_dir
+    @skip_if_lt_x_gpu(2)
+    def test_1d_to_2d_reshard_mesh_change(self) -> None:
+        CHECKPOINT_DIR = self.temp_dir
+        for placements_1d in ONE_D_PLACEMENTS:
+            global_tensor = torch.arange(16, dtype=torch.float).view(4, 4)
+            mesh_shape = (self.world_size,)
+            mesh_1d = init_device_mesh(self.device_type, mesh_shape)
+            dtensor = distribute_tensor(
+                global_tensor, mesh_1d, placements=placements_1d
+            )
+            state_dict_to_save = {"dtensor": dtensor}
+
+            dist_cp.save_state_dict(
+                state_dict=state_dict_to_save,
+                storage_writer=dist_cp.FileSystemWriter(path=CHECKPOINT_DIR),
+                planner=dist_cp.DefaultSavePlanner(),
+            )
```

Comment:
lots of codes here seems to be duplicate. Maybe you want to consolidate them a little bit?